apps/vmq_server/src/vmq_ranch.erl

```diff
@@ -177,14 +177,15 @@ loop_({exit, Reason, State}) ->
     _ = internal_flush(State),
     teardown(State, Reason).
 
-teardown(#st{socket = Socket}, Reason) ->
+teardown(#st{socket = Socket, fsm_mod = FsmMod, fsm_state = FsmState}, Reason) ->
     case Reason of
         normal ->
             lager:debug("session normally stopped", []);
         shutdown ->
             lager:debug("session stopped due to shutdown", []);
         _ ->
-            lager:warning("session stopped abnormally due to '~p'", [Reason])
+            SubscriberId = apply(FsmMod, subscriber, [FsmState]),
```

Comment:
FsmState should already carry the subscriber id here. Am I missing a point? (reason for the function call)

Reply:
That is true, but there are three different state records from the three fsms that could appear here (mqtt, mqtt5, mqtt_pre), so the actual record used depends on the FsmMod. I am still an Erlang beginner, but it seems Erlang does not allow one to reference a record (or export it's definition) that is defined in another module. I found only two ways of implementing it, one the function call (which is called only in an error case, so it should be ok performance wise) or moving the record definition into header files and reference them directly. Keeping the record module-locale and encapsulate it via a function seemed more natural to me. Let me know if you prefer the header idea, or if there is a third way I do not know about yet.